executor/executor.go

```diff
@@ -419,56 +424,128 @@ func (e *CheckTableExec) Open(ctx context.Context) error {
 	}
-	defer func() { e.done = true }()
-	for _, t := range e.tables {
-		dbName := t.DBInfo.Name
-		tb, err := e.is.TableByName(dbName, t.Name)
+
+	close(e.exitCh)
+	var err error
+	for _, src := range e.srcs {
+		err1 := src.Close()
+		if err == nil && err1 != nil {
+			err = err1
+		}
+	}
+	return errors.Trace(err)
+}
+
+func (e *CheckTableExec) checkIndexHandle(ctx context.Context, num int, src *IndexLookUpExecutor) error {
+	cols := src.schema.Columns
+	retFieldTypes := make([]*types.FieldType, len(cols))
+	for i := range cols {
+		retFieldTypes[i] = cols[i].RetType
+	}
+	chk := chunk.New(retFieldTypes, e.initCap, e.maxChunkSize)
+	iter := chunk.NewIterator4Chunk(chk)
+
+	var err error
+	handles := make(map[int64]struct{}, 1024)
+	for {
+		err = src.Next(ctx, chk)
 		if err != nil {
-			return errors.Trace(err)
+			break
 		}
-		if tb.Meta().GetPartitionInfo() != nil {
-			err = e.doCheckPartitionedTable(tb.(table.PartitionedTable))
-		} else {
-			err = e.doCheckTable(tb)
+		if chk.NumRows() == 0 {
+			break
 		}
-		if err != nil {
-			log.Warnf("%v error:%v", t.Name, errors.ErrorStack(err))
-			if admin.ErrDataInConsistent.Equal(err) {
-				return ErrAdminCheckTable.GenWithStack("%v err:%v", t.Name, err)
+
+		for chunkRow := iter.Begin(); chunkRow != iter.End(); chunkRow = iter.Next() {
+			datum := chunkRow.GetDatum(0, retFieldTypes[0])
+			h := datum.GetInt64()
+			if _, ok := handles[h]; ok {
+				err = errors.Errorf("admin check table %s, index %s handle %d more than one", e.tblInfo.Name, e.indices[num].Meta().Name, h)
+				e.retCh <- errors.Trace(err)
+				return errors.Trace(err)
 			}
+			handles[h] = struct{}{}
+		}
 
-			return errors.Errorf("%v err:%v", t.Name, err)
+		select {
+		case <-e.exitCh:
+			return nil
+		default:
 		}
 	}
-	return nil
+	e.retCh <- errors.Trace(err)
+	return errors.Trace(err)
 }
 
-func (e *CheckTableExec) doCheckPartitionedTable(tbl table.PartitionedTable) error {
-	info := tbl.Meta().GetPartitionInfo()
-	for _, def := range info.Definitions {
-		pid := def.ID
-		partition := tbl.GetPartition(pid)
-		if err := e.doCheckTable(partition); err != nil {
+// Next implements the Executor Next interface.
+func (e *CheckTableExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+	if e.done || len(e.srcs) == 0 {
+		return nil
+	}
+	defer func() { e.done = true }()
+
+	idxNames := make([]string, 0, len(e.indices))
+	for _, idx := range e.indices {
+		idxNames = append(idxNames, idx.Meta().Name.O)
+	}
+	greater, idxOffset, err := admin.CheckIndicesCount(e.ctx, e.dbName, e.tblInfo.Name.O, idxNames)
+	if err != nil {
+		log.Warnf("check table %v, greater %v index %s err: %v", e.tblInfo.Name, greater, idxNames[idxOffset], errors.ErrorStack(err))
+		tbl := e.srcs[idxOffset].table
+		txn := e.ctx.Txn(true)
+		switch greater {
+		case admin.IdxCntGreater:
+			err = e.checkIndexHandle(ctx, idxOffset, e.srcs[idxOffset])
+		case admin.TblCntGreater:
+			err = e.checkTableRecord(txn, tbl, e.indices[idxOffset])
```

Comment:
If the count of rows on the table is greater, the `admin check table` will run as before?

Reply:
No. Now we only call `CheckRecordAndIndex` here. Before this PR, we call `checkIndexAndRecord` and `CheckRecordAndIndex`.